docs/sphinx/examples/python/qmcmc.py

```diff
@@ -0,0 +1,184 @@
+# To run with the default values of the parameters:
+# python qmcmc.py
+# 
+# For non-default values of the parameters:
+# python3 qmcmc.py --num_iterations 30 --nqubits 13 --temperature 0.08 --shots_count 20
+# 
+# This code is based on the quantum enchanced Markov 
+# Chain Monte Carlo (QMCMC) algorithm presented in the
+# paper https://arxiv.org/pdf/2203.12497.pdf
+# The Hamiltonian here is exponentiated using the first
+# order Trotterization.
+
+import cudaq
+from cudaq import spin
+import numpy as np
+import random
+import argparse
+
+# Generate a random bitstring of given length
+def generate_random_bitstring(length):
+    return ''.join(random.choice('01') for _ in range(length))
+
+# Initialize the circuit with a given bitstring
+def initial_state(qc, q, bitstring):
+    for i, bit in enumerate(bitstring):
+        if bit == '1':
+            qc.x(q[i])
+    return qc
+
+# Calculate energy for a spin configuration 
+def calculate_energy(bitstring, J, h):
+    # map the bitstring to the spin configuration first
+    nqubits = len(bitstring)
+    s = [-1 if bit == '1' else 1 for bit in bitstring]
+    sum_E = 0.0
+    for k in range(1, nqubits):
+        for j in range(k+1, nqubits):
+            sum_E -= J[j][k]*int(s[j])*int(s[k])
+
+    for j in range(1, nqubits):
+        sum_E -= h[j]*int(s[j])
+
+    return sum_E
+
+# Create a list of Hamiltonians if order matters
+def generate_H(gamma, nqubits, J, h):
+    # Initialize Hamiltonians with 0 coefficients
+    H_prob = 0 * spin.i(0) 
+    H_mix = 0 * spin.i(0) 
+    H_list = []
+    count_problem_terms = 0
+
+    # problem hamiltonian  
+    for k in range (1, nqubits):
+        for j in range(k+1, nqubits):
+            H_prob -= J[j][k]* spin.z(j)*spin.z(k)
+            H_list.append((gamma - 1.0) * J[j][k]* spin.z(j)*spin.z(k))
+            count_problem_terms = count_problem_terms + 1
+
+    for j in range(1, nqubits):
+        H_prob -= h[j]*spin.z(j)
+        H_list.append((gamma - 1.0) * h[j]*spin.z(j))
+        count_problem_terms =  count_problem_terms + 1
+
+    # mixer hamiltonian
+    for j in range(1, nqubits):
+        H_mix += spin.x(j)
+        H_list.append(gamma * spin.x(j))
+
+    # final hamiltonian
+    alpha = np.linalg.norm(H_mix.to_matrix()) / np.linalg.norm(H_prob.to_matrix())
+    for i in range(count_problem_terms):
+        H_list[i] = H_list[i] * alpha
+
+    ordered_H = []
+    for term in H_list:
+        adjTerm = term
+        for i in range(term.get_qubit_count(), nqubits):
+            adjTerm = term * spin.i(i)
+        ordered_H.append(adjTerm)
+
+    return ordered_H
+
+# This is the first-order Trotter gate decomposition
+def trotter_circuit(kernel, qreg, hk, dt, n_qubits):
+    for term in hk:
+        qbitSupport = []
```

Comment:
With a simple additional binding in `py_kernel_builder.cpp` we could replace this loop body with 
```python
for term in hk:
    kernel.exp_pauli(2. * dt * term.get_coefficient().real, qreg, term)
return kernel
```
(You could also just use the existing binding with your string split `pauliString` line) 
```cpp
.def(
          "exp_pauli",
          [](kernel_builder<> &self, py::object theta, const QuakeValue &qubits,
             spin_op &pauli) {
            if (pauli.num_terms() != 1)
              throw std::runtime_error(
                  "exp_pauli expects a SpinOperator with a single term.");

            if (py::isinstance<py::float_>(theta))
              self.exp_pauli(theta.cast<double>(), qubits,
                             pauli.to_string(/*printCoeff*/ false));
            else if (py::isinstance<QuakeValue>(theta))
              self.exp_pauli(theta.cast<QuakeValue &>(), qubits,
                             pauli.to_string(/*printCoeff*/ false));
            else
              throw std::runtime_error(
                  "Invalid `theta` argument type. Must be a "
                  "`float` or a `QuakeValue`.");
          },
          "Apply a general Pauli tensor product rotation, `exp(i theta P)`, on "
          "the specified qubit register. The Pauli tensor product is provided "
          "as a `SpinOperator` with a single term, e.g. `XXYX` for a 4-qubit "
          "term. The angle parameter "
          "can be provided as a concrete float or a `QuakeValue`. The "
          "coefficient of the `SpinOperator` is ignored (user must multiply "
          "the coefficient with the `theta` value explicitly).")
```

Reply:
Okay, sounds good - thanks!